airmail_indexer/src/osm_pbf.rs

```diff
@@ -0,0 +1,182 @@
+                }
+
+                // A way is a polyline or polygon, like a road.
+                Element::Way(way) => {
+                    // Attempt to get the location from the way from the underlying way data,
+                    // this requires the way is stored with the option LocationsOnWays enabled.
+                    let mut location = way
+                        .node_locations()
+                        .next()
+                        .map(|location| (location.lat(), location.lon()));
+
+                    // If the location is not present in the way data, attempt to get the location
+                    // from the node_map we previously built.
+                    if location.is_none() {
+                        location = way
+                            .refs()
+                            .next()
+                            .and_then(|node_id| {
+                                self.indexer_cache.query_node_location(node_id).ok()
+                            })
+                            .flatten();
+                    }
+
+                    if let Some(location) = location {
+                        let tags = way
+                            .tags()
+                            .collect::<HashMap<_, _>>();
+                        if let Some(interesting_poi) =
+                            OsmPoi::new(tags, location).and_then(OsmPoi::index_poi)
+                        {
+                            count_ways.fetch_add(1, Ordering::Relaxed);
+                            self.sender.send(interesting_poi).expect("sender failed");
+                            1
+                        } else {
+                            0
+                        }
+                    } else {
+                        0
+                    }
+                }
+
+                // Ignored
+                Element::Relation(_) |
+
+                // Seems to be unused, as dense node is used instead
+                Element::Node(_) => 0,
+            },
+            || 0_u64,
+            |a, b| a + b,
+        )?;
+
+        let count_ways = count_ways.load(Ordering::Relaxed);
+        let count_dense_nodes = count_dense_nodes.load(Ordering::Relaxed);
+
+        info!(
+            "Loaded {} interesting pois, made up of {} dense nodes,  and {} ways",
+            pois, count_dense_nodes, count_ways
+        );
+
+        if count_ways == 0 {
+            warn!("No ways found in OSM PBF file. Ensure your pbf file has locations present, see Google: LocationsOnWays");
+        }
+
+        Ok(())
+    }
+
+    fn cache_nodes_for_ways(&self) -> Result<()> {
+        // Increase buffer to reduce writes to disk
+        self.indexer_cache.buffer_size(10_000_000)?;
+
+        // Store a map of nodes to their locations for quick lookups in the second pass.
+        let cached_count = ElementReader::from_path(&self.pbf_path)?.par_map_reduce(
+            |element| match element {
+                Element::DenseNode(node) => {
```

Comment:
Same as above, I'd prefer to handle the "normal" node elements here even if they don't come up in practice.